### PR TITLE
Term entry - PyTorch - nn - Loss Functions

### LINKED ISSUE #5233: [Term Entry] PyTorch nn Loss Functions
### Reason/inspiration (optional)

We would like a new entry on the `Loss Functions` in PyTorch. The entry should go in a new file under `content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md`.      _x000D_
  _x000D_
The entry should include:  _x000D_
  _x000D_
- An introduction to the concept     _x000D_
  _x000D_
- A `Syntax` section that provides the syntax for the concept (if applicable)     _x000D_
  _x000D_
- An `Example` section that provides an example demonstrating the concept in use (if applicable)     _x000D_
  _x000D_
Please refer to the [concept entry template](https://github.com/Codecademy/docs/blob/main/documentation/concept-entry-template.md), [content standards](https://github.com/Codecademy/docs/blob/main/documentation/content-standards.md), and [markdown style guide](https://curriculum-documentation.codecademy.com/content-guidelines/markdown-style-guide/) when working on the PR for this issue.

### Entry Type (select all that apply)

New Term for Existing Concept Entry

### Code of Conduct

- [X] By submitting this issue, I agree to follow Codecademy Doc's [Code of Conduct](https://github.com/Codecademy/docs/blob/main/.github/CODE_OF_CONDUCT.md).

### For Maintainers

- [X] Labels added
- [ ] Issue is assigned

### PR DESCRIPTION
**Pull request by danitellini**
### Description

Topic: PyTorch's nn `Loss Functions`
File created: `docs/content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md`
Completed for new opportunity to learn

### Issue Solved

Closes #5233 

### Type of Change

- Adding a new entry
- ~Editing an existing entry (fixing a typo, bug, issues, etc)~
- ~Updating the documentation~

### Checklist

- [x] All writings are my own.
- [x] My entry follows the Codecademy Docs style guide.
- [x] My changes generate no new warnings.
- [x] I have performed a self-review of my own writing and code.
- [x] I have checked my entry and corrected any misspellings.
- [x] I have made corresponding changes to the documentation if needed.
- [x] I have confirmed my changes are not being pushed from my forked `main` branch.
- [x] I have confirmed that I'm pushing from a new branch named after the changes I'm making.
- [x] I have linked any issues that are relevant to this PR in the `Issues Solved` section.

### CODE REVIEW

**Comment by danitellini** [comment]
I had a hard time writing this. If this isn't enough, please let me know what I need to add. I haven't done an entry on a term that led to others 😅

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 1]
```diff
@@ -0,0 +1,230 @@
+---
+Title: 'Loss Functions'
+Description: 'Represents a multi-dimensional array used for data storage and computation in PyTorch.'
+Subjects:
+  - 'Machine Learning'
+  - 'Python'
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:6]: `Python` is not a valid subject. For more subjects, please refer to [subjects.md](https://github.com/Codecademy/docs/blob/main/documentation/subjects.md). Here are some suggestions -

```suggestion
  - 'Data Science'
  - 'AI'
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 2]
```diff
@@ -0,0 +1,230 @@
...
+Tags:
+  - 'PyTorch'
+  - 'Tensors'
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:9]: `Tensors` is not a valid tag. For valid tags, please refer [tags.md](https://github.com/Codecademy/docs/blob/main/documentation/tags.md).

```suggestion
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 3]
```diff
@@ -0,0 +1,230 @@
...
+  - 'Deep Learning'
+  - 'Data Structures'
+CatalogContent:
+  - 'learn-python-3'
+  - 'paths/machine-learning'
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:14]: We can add more PyTorch-specific catalog content here. Please refer to [catalog-content.md](https://github.com/Codecademy/docs/blob/main/documentation/catalog-content.md). Here are some suggestions:

```suggestion
  - 'learn-python-3'
  - 'intro-to-py-torch-and-neural-networks'
  - 'py-torch-for-classification'
  - 'paths/machine-learning'
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 4]
```diff
@@ -0,0 +1,230 @@
...
+---
+
+**`Loss Functions`**, also known as `cost functions`, are critical in [neural network](https://github.com/Codecademy/docs/blob/main/content/pytorch/concepts/nn/nn.md) training. They quanitfy how well the model's predictions align with the target values. By computing a scalar loss value, they enable optimization algorithms to adjust the model's weights to improve performance.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:17]: Since 'cost functions' is a new term and not the exact keyword in the code, we should only emphasize it in bold, not backticks. Also, we should link the published doc link rather than using the GitHub link here.

```suggestion
**`Loss Functions`**, also known as **cost functions**, are critical in [neural network](https://www.codecademy.com/resources/docs/pytorch/nn) training. 

They quantify how well the model's predictions align with the target values. 

By computing a scalar loss value, they enable optimization algorithms to adjust the model's weights to improve performance.
```

Comment (PragatiVerma18) [reply content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:17]: Also, the introduction(first line) should be mostly the same as the description. Please update accordingly.

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 5]
```diff
@@ -0,0 +1,230 @@
...
+
+[PyTorch's](https://github.com/Codecademy/docs/blob/main/content/pytorch/pytorch.md) `nn` (neural network) module offers various built-in loss functions suitable for different tasks, such as regression and classification.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:19]: Linking is not required here as it will be published inside the PyTorch concepts, so it can be easily accessed using UI once published.

```suggestion
PyTorch's `nn` (neural network) module offers various built-in loss functions suitable for different tasks, such as regression and classification.
```

Comment (PragatiVerma18) [reply content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:19]: This is not required as we have already linked to the specific doc entry that explains what is `nn` module.

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 6]
```diff
@@ -0,0 +1,230 @@
...
+# Define the loss function.
+criterion = nn.LossFunctionName(*args, **kwargs)
+
+# Compute the loss.
+loss = criterion(predicted_outputs, target_values)
+```
+
+- `LossFunctionName` = Replace with the specific loss function (e.g. `MSELoss` which is used for regression, `CrossEntropyLoss` which is used for classification).
+- `predicted_outputs` = The predicted outputs from your model.
+- `target_values` = The real data values.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:37]: We typically use `:` rather than `=` when explaining list items. Also, we should `,` in first sentence to make it more readable.

```suggestion
- `LossFunctionName`: Replace with the specific loss function (e.g. `MSELoss`, which is used for regression, `CrossEntropyLoss`, which is used for classification).
- `predicted_outputs`: The predicted outputs from your model.
- `target_values`: The real data values.
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 7]
```diff
@@ -0,0 +1,230 @@
...
+
+# Backward pass and optimization.
+optimizer.zero_grad()
+loss.backward()
+optimizer.step()
+
+# Forward pass after optimization.
+outputs = model(inputs)
+loss = criterion(outputs, targets)
+print('Loss after one optimization step:', loss.item())
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:82]: Since the comments are not complete sentences, we can skip adding the full stop at the end.

```suggestion
import torch
import torch.nn as nn
import torch.optim as optim

# Hyperparameters
input_size = 10
output_size = 1
learning_rate = 0.01
batch_size = 32

# Sample data
inputs = torch.randn(batch_size, input_size)
targets = torch.randn(batch_size, output_size)

# Define a simple linear regression model
model = nn.Linear(input_size, output_size)

# Define the loss function
criterion = nn.MSELoss()

# Define the optimizer
optimizer = optim.SGD(model.parameters(), lr=learning_rate)

# Forward pass
outputs = model(inputs)

# Compute Loss
loss = criterion(outputs, targets)
print('Initial Loss:', loss.item())

# Backward pass and optimization
optimizer.zero_grad()
loss.backward()
optimizer.step()

# Forward pass after optimization
outputs = model(inputs)
loss = criterion(outputs, targets)
print('Loss after one optimization step:', loss.item())
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 8]
```diff
@@ -0,0 +1,230 @@
...
+```
+
+Sample output for this example:
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:85]: We should rewrite it for better understanding.

```suggestion
The output for this example will be as follows:
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 9]
```diff
@@ -0,0 +1,230 @@
...
+loss.backward()
+optimizer.step()
+
+# Forward pass after optimization
+outputs = model(inputs)
+loss = criterion(outputs, targets)
+print('Loss after one optimization step:', loss.item())
+```
+
+Sample output for this example:
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:143]: Rewrite it for better understanding:

```suggestion
The output for this example will be as follows:
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 10]
```diff
@@ -0,0 +1,230 @@
...
+loss.backward()
+optimizer.step()
+
+# Forward pass after optimization
+outputs = model(inputs)
+loss = criterion(outputs, targets)
+print('Loss after one optimization step:', loss.item())
+```
+
+Sample output for the example:
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:200]: Rewrite for better understanding:

```suggestion
The output for this example will be as follows:
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 11]
```diff
@@ -0,0 +1,230 @@
...
+- `nn.MSELoss`: Mean Squared Error Loss. Suitable for regression tasks.
+- `nn.CrossEntropyLoss`: Combines `nn.LogSoftmax` and `nn.NLLLoss`. Used for multi-class classification.
+- `nn.NLLLoss`: Negative Log Likelihood Loss. Often used with `nn.LogSoftmax`.
+- `nn.BCELoss`: Binary Cross Entropy Loss. Used for binary classification tasks.
+- `nn.BCEWithLogitsLoss`: Combines a **Sigmoid** layer and the BCELoss. More numerically stable.
+- `nn.L1Loss`: Mean Absolute Error Loss.
+- `nn.SmoothL1Loss`: A combination of `nn.L1Loss` and `nn.MSELoss`. Less sensitive to outliers.
+- `nn.HingeEmbeddingLoss`: Used for measuring similarity between samples.
+
+_Sigmoid function_ is a mathematical function that has an "S" shaped curve (sigmoidal shape). It is widely used in machine learning, particularly in logistic regression and neural networks, as an activation function to introduce non-linearity into the model and to map any real-valued number into a value between 0 and 1.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:222]: We should skip explaining common loss functions here. This will make the entry longer, and beginners will not get an exact idea of what the function does.

These can be separately covered as term entries.


```suggestion
```

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 12]
```diff
@@ -0,0 +1,230 @@
...
+
+## Notes
+
+- Input Requirements: Ensure that the shapes and data types of `predicted_outputs` and `target_values` match the expectations of the loss function.
+  - For `nn.CrossEntropyLoss`, `predicted_outputs` should contain raw scores (logits), and `target_values` should contain class indices (not one-hot encoded).
+  - For `nn.BCEWithLogitsLoss`, `predicted_values` should be raw scores; the loss function applies the Sigmoid function internally.
+- Reduction Parameter: Most loss functions have a `reduction` parameter (`mean`, `sum`, or `none`) that specifies how the loss should be aggregated.
+- Class Imbalance: For datasets with class imbalance, consider using the `weight` parameter in the loss function to assign different weights to classes.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:230]: We should avoid such long notes. Try to accommodate them in paragraph form where we are explaining the syntax itself.

**Review by PragatiVerma18 — CHANGES_REQUESTED** [review]
Hey @danitellini, thanks for adding this entry, I have some suggestions. Please check the comments and do the required changes. Let me know if you have any questions.

**Comment by danitellini** [comment]
All set! @PragatiVerma18

File: content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md [hunk 13]
```diff
@@ -0,0 +1,215 @@
...
+  - 'Data Structures'
+CatalogContent:
+  - 'intro-to-py-torch-and-neural-networks'
+  - 'py-torch-for-classification'
+  - 'paths/machine-learning'
+---
+
+**`Loss Functions`**, also known as **cost functions**, represent critical functions used to optimize [neural network](https://www.codecademy.com/resources/docs/pytorch/nn) predictions in PyTorch.
+
+They quanitfy how well the model's predictions align with the target values.
```
Comment (PragatiVerma18) [inline content/pytorch/concepts/nn/terms/loss-functions/loss-functions.md:19]:
```suggestion
They quantify how well the model's predictions align with the target values.
```

**Review by PragatiVerma18 — APPROVED** [review]
Hey @danitellini, thanks for getting back with the changes so quickly. This PR is now ready for a second review. 🚀

**Review by mamtawardhani — APPROVED** [review]
Thank you for contributing to Codecademy Docs @danitellini 😄 

The entry looks good to be merged! 🚀
